bundlewrap/operations.py

```diff
@@ -249,6 +256,45 @@ def run(
     return result
 
 
+def run_routeros(hostname, username, password, *args):
+    with ROUTEROS_CONNECTIONS_LOCK:
+        try:
+            connection, lock = ROUTEROS_CONNECTIONS[hostname]
+        except KeyError:
+            try:
+                connection = connect(
+                    # str() to resolve Faults
+                    username=str(username),
+                    password=str(password or ""),
+                    host=hostname,
+                    timeout=120.0,
+                )
+                lock = Lock()
+                ROUTEROS_CONNECTIONS[hostname] = connection, lock
+            except Exception as e:
+                raise RemoteException(str(e)) from e
+
+    try:
+        io.debug(f'{hostname}: running routeros command: {repr(args)}')
+        with lock:
```

Comment:
What about this scenario:

1.  Thread A grabs this lock and runs `rawCmd()`.
2.  Thread B wants to do the same and waits at grabbing the lock.
3.  Thread A runs into an error, let's say it's a timeout (this is the most harmful situation, because the connection is still "working" but data is pending; this is what we originally saw). Thread A thus releases the lock.
4.  Thread B grabs the lock and tries to run `rawCmd()`, but the connection is in an undefined state now.
5.  Thread A only now proceeds to run `connection.close()`, but it's too late now, thread B already did its thing.

Should `with lock` be move to the outermost level?

Reply:
Yes, but that's still not enough. Had to decouple both locks even more to avoid a deadlock.